error.go

```diff
@@ -18,68 +18,69 @@ package edgedb
 
 import (
 	"errors"
-	"fmt"
-	"log"
 
 	"github.com/edgedb/edgedb-go/internal/buff"
 )
 
-var (
-	// Error is wrapped by all edgedb errors.
-	Error error = errors.New("")
+var errZeroResults error = &noDataError{msg: "zero results"}
 
-	// ErrReleasedTwice is returned if a PoolConn is released more than once.
-	ErrReleasedTwice = fmt.Errorf(
-		"connection released more than once%w", Error,
-	)
+// Error is wrapped by all errors.
+type Error interface {
+	Error() string
+	isError()
+}
 
-	// ErrZeroResults is returned when a query has no results.
-	ErrZeroResults = fmt.Errorf("zero results%w", Error)
+type baseError struct {
+	msg string
+	err error
+}
 
-	// ErrPoolClosed is returned by operations on closed pools.
-	ErrPoolClosed error = fmt.Errorf("pool closed%w", Error)
+func (e *baseError) isError() {}
 
-	// ErrContextExpired is returned when an expired context is used.
-	ErrContextExpired error = fmt.Errorf("context expired%w", Error)
+func (e *baseError) Error() string {
+	if e.err != nil {
+		return e.err.Error()
+	}
 
-	// ErrBadConfig is wrapped
-	// when a function returning Options encounters an error.
-	ErrBadConfig error = fmt.Errorf("%w", Error)
+	return e.msg
+}
 
-	// ErrClientFault ...
-	ErrClientFault error = fmt.Errorf("%w", Error)
+func (e *baseError) Unwrap() error {
+	return e.err
+}
 
-	// ErrInterfaceViolation ...
-	ErrInterfaceViolation error = fmt.Errorf("%w", ErrClientFault)
-)
+func firstError(a, b error) error {
+	if a != nil {
+		return a
+	}
 
-func decodeError(r *buff.Reader) error {
-	r.Discard(5) // skip severity & code
-	err := fmt.Errorf("%v%w", r.PopString(), Error)
+	return b
+}
 
+func decodeError(r *buff.Reader) error {
+	r.Discard(1) // severity
+	err := errorFromCode(r.PopUint32(), r.PopString())
 	n := int(r.PopUint16())
-	headers := make(map[uint16]string, n)
 
 	for i := 0; i < n; i++ {
-		headers[r.PopUint16()] = r.PopString()
+		r.PopUint16() // key
+		r.PopString() // value
 	}
 
-	// todo do something with headers
-	log.Println(headers)
 	return err
 }
 
 type wrappedManyError struct {
-	msg     string
-	wrapped []error
+	msg  string
```

Comment:
I don't understand what this struct is needed for...

Reply:
see https://github.com/edgedb/edgedb-go/pull/51#discussion_r553584300